### PR TITLE
perf: add dLLM consumer-sufficient TP vocab state

### PR DESCRIPTION
**Pull request by peareji-duck**
## Motivation

This PR adds and hardens a dLLM TP-local vocab state path for SGLang.

For LowConfidence dLLM inference, the logits processor can pass compact consumer-sufficient vocab state instead of materializing / gathering full logits across tensor-parallel ranks. This reduces unnecessary full-vocab communication when the downstream dLLM consumer only needs argmax and confidence state.

This PR also fixes a prefill CUDA graph replay bug for dLLM outputs where `next_token_logits=None` is valid but was sliced unconditionally.

## Modifications

| Area | Change |
| --- | --- |
| Logits processor | Add compact `dllm_vocab_state` path for TP-local vocab consumers |
| TP communication | Avoid full-vocab dense state when compact state is enough |
| dLLM algorithms | Consume compact argmax / confidence state in LowConfidence-compatible paths |
| Prefill CUDA graph | Slice dLLM `full_logits` / `dllm_vocab_state` safely while preserving `next_token_logits=None` |
| Guarding | Default-off env-gated path with dense fallback where compact state is insufficient |
| Tests | Add regression coverage for compact vocab state, packed gather, dense equivalence, endpoint comparison hook, and CUDA graph slicing |

## Accuracy Tests

Focused regression tests passed:

```text
25 passed, 1 skipped, 21 warnings
```

The skipped test is an endpoint-comparison gate that requires already-running baseline and optimized endpoint URLs.

The compact vocab state path is tested against dense logits for argmax and confidence consumers, including padded vocab shards, packed gather, tie-breaking, and logit-scale / softcap ordering. The CUDA graph fix has regression coverage for dLLM outputs with `next_token_logits=None`.

This benchmark is a serving-throughput benchmark, not a downstream task-accuracy evaluation.

## Speed Tests and Profiling

Measured through SGLang endpoints using `vllm bench serve`, TP4, closed-loop load, random input length 30, output length 32, 512 measured prompts, 256 warmups.

Benchmark setup:

| item | value |
| --- | --- |
| server | `python -m sglang.launch_server` OpenAI-compatible endpoint |
| tensor parallelism | TP4 |
| load | closed loop, `--request-rate inf` |
| prompts | 512 measured prompts, 256 warmup prompts |
| random shape | input length 30, output length 32 |
| concurrency | c4, c8, c16, c32 |
| baseline flag | `SGLANG_DLLM_TP_LOCAL_VOCAB=0` |
| optimized flag | `SGLANG_DLLM_TP_LOCAL_VOCAB=1` |
| packed gather flag | `SGLANG_DLLM_TP_LOCAL_VOCAB_PACKED_GATHER=1` |
| algorithm | `LowConfidence` |
| attention / sampling backend | `flashinfer` |

Models:

| family | model |
| --- | --- |
| LLaDA2-flash | `inclusionAI/LLaDA2.0-flash` |
| SDAR-8B | `JetLM/SDAR-8B-Chat` |
| SDAR-30B-A3B | `JetLM/SDAR-30B-A3B-Chat-b32` |

Results:

| model | claim range | result |
| --- | --- | ---: |
| LLaDA2-flash | c8-c32 geomean | 1.049x |
| SDAR-8B | c8-c32 geomean | 1.076x |
| SDAR-30B-A3B | repeated c16/c32 geomean | 1.054x |

Detailed rows:

| model | c4 | c8 | c16 | c32 |
| --- | ---: | ---: | ---: | ---: |
| LLaDA2-flash | 1.014x | 1.030x | 1.056x | 1.061x |
| SDAR-8B | 1.027x | 1.040x | 1.098x | 1.090x |
| SDAR-30B-A3B, repeat n=3 | - | - | 1.046x | 1.061x |

An initial SDAR-30B c32 run showed `1.433x`, but repeat validation did not reproduce it. The repeated c16/c32 row above is the PR claim.

Reproduction:

```bash
export MODEL=JetLM/SDAR-8B-Chat

# baseline endpoint
SGLANG_DLLM_TP_LOCAL_VOCAB=0 \
SGLANG_DLLM_TP_LOCAL_VOCAB_PACKED_GATHER=1 \
python -m sglang.launch_server \
  --model-path "${MODEL}" \
  --trust-remote-code \
  --host 127.0.0.1 \
  --port 8000 \
  --tp-size 4 \
  --dllm-algorithm LowConfidence \
  --attention-backend flashinfer \
  --sampling-backend flashinfer \
  --max-running-requests 32

# optimized endpoint
SGLANG_DLLM_TP_LOCAL_VOCAB=1 \
SGLANG_DLLM_TP_LOCAL_VOCAB_PACKED_GATHER=1 \
python -m sglang.launch_server \
  --model-path "${MODEL}" \
  --trust-remote-code \
  --host 127.0.0.1 \
  --port 8001 \
  --tp-size 4 \
  --dllm-algorithm LowConfidence \
  --attention-backend flashinfer \
  --sampling-backend flashinfer \
  --max-running-requests 32
```

Use the same command shape for `inclusionAI/LLaDA2.0-flash` and
`JetLM/SDAR-30B-A3B-Chat-b32`. The measured LLaDA2-flash and SDAR-30B-A3B
runs additionally used `--cuda-graph-backend-decode=disabled`.

Benchmark both endpoints:

```bash
for PORT in 8000 8001; do
  for C in 4 8 16 32; do
    vllm bench serve \
      --backend openai \
      --base-url "http://127.0.0.1:${PORT}" \
      --endpoint /v1/completions \
      --model "${MODEL}" \
      --served-model-name "${MODEL}" \
      --tokenizer "${MODEL}" \
      --trust-remote-code \
      --dataset-name random \
      --random-input-len 30 \
      --random-output-len 32 \
      --random-range-ratio 0 \
      --num-prompts 512 \
      --num-warmups 256 \
      --request-rate inf \
      --max-concurrency "${C}" \
      --ignore-eos \
      --temperature 0 \
      --seed "$((100000 + C))" \
      --percentile-metrics ttft,tpot,itl,e2el \
      --metric-percentiles 50,90,95,99
  done
done
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Comment by Kokoro2336** [comment]
@peareji-duck Hey, are you still working on this issue? If not, I'd like to take this and refactor the code, since the changes are stale.
